### PR TITLE
Add tickles sleep for STM32F1 CPU

### PR DESCRIPTION
**Pull request by jipanienko**
Tickles sleep can be used when LSE oscillator is used and OS_TICKS_PER_SEC is set to power of 2.

### CODE REVIEW

File: hw/mcu/stm/stm32_common/syscfg.yml [hunk 1]
```diff
@@ -262,3 +262,7 @@ syscfg.defs:
     ADC_2:
         description: "ADC_2"
         value:  0
+
+    OS_TICKS_PER_SEC:
+        description:
```
Comment (kasjer) [inline hw/mcu/stm/stm32_common/syscfg.yml:267]: Description not provided.

File: hw/mcu/stm/stm32f1xx/src/system_stm32f1xx.c [hunk 2]
```diff
@@ -69,6 +69,12 @@ void SystemClock_Config(void);
 void
 SystemInit(void)
 {
+    /* Reset all peripherals. */
+    RCC->APB1RSTR = 0xFFFFFFFF;
+    RCC->APB2RSTR = 0xFFFFFFFF;
+    RCC->APB1RSTR = 0x00000000;
+    RCC->APB2RSTR = 0x00000000;
+
```
Comment (kasjer) [inline hw/mcu/stm/stm32f1xx/src/system_stm32f1xx.c:77]: This code does not seem to be relevant to the change, isn't it?

File: hw/mcu/stm/stm32_common/src/hal_os_tick.c [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #include <assert.h>
 #include "os/mynewt.h"
 #include <hal/hal_os_tick.h>
+#include <stm32_common/stm32_hal.h>
 
 /*
  * XXX implement tickless mode.
```
Comment (kasjer) [inline hw/mcu/stm/stm32_common/src/hal_os_tick.c:26]: PR seems to make this comment partially obsolete :)

File: hw/mcu/stm/stm32_common/src/hal_os_tick.c [hunk 4]
```diff
@@ -76,3 +79,111 @@ os_tick_init(uint32_t os_ticks_per_sec, int prio)
...
+    }
+}
+
+void
+os_tick_init(uint32_t os_ticks_per_sec, int prio)
+{
+    uint32_t sr;
+
+    __HAL_RCC_RTC_CONFIG(RCC_RTCCLKSOURCE_LSE);
+    __HAL_RCC_RTC_ENABLE();
```
Comment (kasjer) [inline hw/mcu/stm/stm32_common/src/hal_os_tick.c:163]: This may not be enough. ST HAL code for changing clock source does a lot of other things since changing it may not be possible after it was first set (for example if bootloader was not configured to use LSE and application is change may not be applied).
Have you checked that it will work in this scenario?
Maybe ST HAL could be used for this?

File: hw/mcu/stm/stm32_common/src/hal_os_tick.c [hunk 5]
```diff
@@ -76,3 +79,111 @@ os_tick_init(uint32_t os_ticks_per_sec, int prio)
     DBGMCU->CR |= (DBGMCU_CR_DBG_STOP | DBGMCU_CR_DBG_STANDBY);
 #endif
 }
+
```
Comment (brianwyld) [inline hw/mcu/stm/stm32_common/src/hal_os_tick.c:82]: IMHO such a large block of STM32F1 specific code should not be the stm32_common files. Should be in the specific MCU directory instead... especially if we then have tickless code for the STM32L1 (as per my PR), STM32L0/L4, etc etc as each one requires a different approach.

**Comment by kasjer** [comment]
Hi @jipanienko, I checked code and it looks ok. I thought I understand how it should work more or less however when I tried this patch to see if it works it did not work at all.
I tried blinky example with following syscfg values that seems to be needed to trigger your code:
```yml
    STM32_CLOCK_LSE: 1
    OS_TICKS_PER_SEC: 128
```
But with just those changes, blinky no longer works.
Am I missing something?

**Comment by jipanienko** [comment]
@kasjer settings are good.
crash is because mcuboot leaves timer running and LSE take time to established in that time interrupt crash because mcuboot memory is no valid and interrupt vectors are not changed yet to blinky or whatever

for workaround start LSE in bootloader STM32_CLOCK_LSE: 1
issue for mcuboot https://github.com/JuulLabs-OSS/mcuboot/issues/759 and https://github.com/JuulLabs-OSS/mcuboot/issues/529

File: hw/mcu/stm/stm32f1xx/src/rtc_tick_stm32f1xx.c [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        if (LL_RTC_IsActiveFlag_SEC(RTC)) {
+            rtc_cnt = (RTC->CNTH << 16u) + RTC->CNTL;
+            ticks--;
+        }
+        rtc_cnt += ticks - 1;
+        /* All flags cleared, disable write protection */
+        RTC->CRL = RTC_CRL_CNF_Msk;
+
+        LL_RTC_ALARM_Set(RTC, rtc_cnt);
+        LL_RTC_EnableIT_ALR(RTC);
```
Comment (kasjer) [inline hw/mcu/stm/stm32f1xx/src/rtc_tick_stm32f1xx.c:83]: I think that SEC interrupt should be disabled otherwise even if alarm is set SEC interrupt will fire with OS_TICKS_PER_SEC frequency.
